File: SudokuSolver/Domain/Cell.cs
using System.ComponentModel.DataAnnotations;

namespace SudokuSolver.Domain;

public class Cell
{
    [Range(1, 9)]
    public int Value { get; set; }
    [Range(0, 8)]
    public int Row { get; }
    [Range(0, 8)]
    public int Column { get; }
    [Range(0, 8)]
    public int Block { get; }
    public HashSet<int> PossibleValues { get; }

    public Cell(int row, int col)
    {
        Value = 0;        
        Row = row;
        Column = col;
        // only the integer part of the number is needed
        Block = 3 * (row / 3) + (col / 3);
        PossibleValues = new() { 1, 2, 3, 4, 5, 6, 7, 8, 9 };
    }

    public void SetValue(int value)
    {
        if (IsOpened())
        {
            throw new ArgumentException($"The Cell({Row},{Column}) has alredy opened and has a value: {Value}. " +
                $"Trying to set another value: {value}.");
        }

        Value = value;
        PossibleValues.Clear();
    }

    public void RemoveFromPossibleValues(int value) => PossibleValues.Remove(value);

    public bool IsOpened() => Value > 0;

    public override string? ToString()
    {
        return $"({Row},{Column},{Block})=>{Value} [{string.Join(',', PossibleValues)}] ";
    }

    public override bool Equals(object? obj)
    {
        return obj is Cell other &&
               Value == other.Value &&
               Row == other.Row &&
               Column == other.Column &&
               Block == other.Block;
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(Value, Row, Column, Block);
    }
}

File: src/main/java/ua/kh/poohdev/sudokusolver/domain/Cell.java
package ua.kh.poohdev.sudokusolver.domain;

import ua.kh.poohdev.sudokusolver.exceptions.InconsistentCellStateException;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import static ua.kh.poohdev.sudokusolver.constants.SudokuSolverConstants.REQUIRED_VALUES;

public class Cell {

    private static final String OPENED_CELL_ERROR_MSG =
        "The Cell (row = %d, col = %d) is already opened and has the value = %d, " +
        "but it has tried to set a new value = %d.";

    private static final String REMOVE_POSSIBLE_VALUE_ERROR_MSG =
        "The Cell (row = %d, col = %d) has NO possible values but still has not opened. " +
        "The last removed possible value = %d.";

    private final int rowNumber;
    private final int colNumber;
    private final int blockNumber;

    private final Set<Integer> possibleValues;
    private int value;

    public Cell(int row, int col) {
        rowNumber = row;
        colNumber = col;
        // only integer part of the number is needed
        blockNumber = 3 * (row / 3) + (col / 3);

        possibleValues = new HashSet<>();
        possibleValues.addAll(REQUIRED_VALUES);

        value = 0;
    }

    public Cell(int rowNumber, int colNumber, int blockNumber, Set<Integer> possibleValues, int value) {
        this.rowNumber = rowNumber;
        this.colNumber = colNumber;
        this.blockNumber = blockNumber;
        this.possibleValues = possibleValues;
        this.value = value;
    }

    public boolean isOpened() {
        return value != 0;
    }

    public void setValue(int newValue) {
        if (isOpened()) {
            throw new InconsistentCellStateException(
                String.format(OPENED_CELL_ERROR_MSG, rowNumber, colNumber, value, newValue));
        }

        value = newValue;
        possibleValues.clear();
    }

    public void removePossibleValue(int value) {
        possibleValues.remove(value);

        if (possibleValues.isEmpty() && !isOpened()) {
            throw new InconsistentCellStateException(
                String.format(REMOVE_POSSIBLE_VALUE_ERROR_MSG, rowNumber, colNumber, value));
        }
    }

    public int getRowNumber() {
        return rowNumber;
    }

    public int getColNumber() {
        return colNumber;
    }

    public int getBlockNumber() {
        return blockNumber;
    }

    public Set<Integer> getPossibleValues() {
        return possibleValues;
    }

    public int getValue() {
        return value;
    }

    public Cell copy() {
        return new Cell(rowNumber, colNumber, blockNumber, new HashSet<>(possibleValues), value);
    }

    @Override
    public String toString() {
//        return "(" + rowNumber + "," + colNumber + "," + blockNumber + ") => " + value +
//               " [" + possibleValues.stream().map(Object::toString).collect(Collectors.joining(",")) + "]";
        return String.valueOf(value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Cell cell = (Cell) o;
        return rowNumber == cell.rowNumber &&
               colNumber == cell.colNumber &&
               blockNumber == cell.blockNumber &&
               value == cell.value;
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowNumber, colNumber, blockNumber, value);
    }
}
